tests/system/action/motion/test_create_sequence_number.py

```diff
@@ -0,0 +1,256 @@
+        self.assertEqual(model.get("sequence_number"), 1)
+
+        response = self.client.post(
+            "/",
+            json=[{"action": "motion.delete", "data": [{"id": 1}]}],
+        )
+        self.assert_status_code(response, 200)
+        model = self.get_model("motion/1")
+        assert model.get("meta_deleted")
+
+        response = self.client.post(
+            "/",
+            json=[
+                {
+                    "action": "motion.create",
+                    "data": [
+                        {
+                            "title": "motion_title",
+                            "meeting_id": 222,
+                            "workflow_id": 12,
+                            "text": "test2",
+                        }
+                    ],
+                }
+            ],
+        )
+        self.assert_status_code(response, 200)
+        model = self.get_model("motion/2")
+        self.assertEqual(model.get("sequence_number"), 2)
+
+    def test_create_sequence_numbers_race_condition(self) -> None:
+        """
+        The main thread in this testing function creates an instance of a threading.lock and acquires it.
+        Thread1 will be started first, but should wait for the acquired lock after reading and before writing.
+        Thread2 will be started later, but passes thread1, because do not acquire the lock.
+        After finishing thread2 the lock will be released in main thread and thread1 could continue with it's writing part.
+
+        The lock-object will be shared in in threading.local(), instance created in shared/lock.py.
+        If possible you should pass as an argument to the thread function(s).
+        """
+        pytest_thread_local.name = "MainThread_RC"
+        self.create_model("meeting/222", {"name": "meeting222"})
+        self.create_workflow(workflow_id=12, meeting_id=222)
+        self.create_workflow(workflow_id=13, meeting_id=222)
+
+        # with MonkeyPatch_DatastoreAdapter_write():
+        with monkeyPatch_DatastoreAdapter_write():
+            testlock = threading.Lock()
+            thread1 = threading.Thread(
+                target=thread_method,
+                kwargs={
+                    "test_instance": self,
+                    "motion_title": "Interrupted motion",
+                    "meeting_id": 222,
+                    "workflow_id": 12,
+                    "testlock": testlock,
+                    "name": "Interrupted Thread",
+                },
+            )
+            thread2 = threading.Thread(
+                target=thread_method,
+                kwargs={
+                    "test_instance": self,
+                    "motion_title": "Passing motion",
+                    "meeting_id": 222,
+                    "workflow_id": 13,
+                    "testlock": None,
+                    "name": "Passing Thread",
+                },
+            )
+
+            testlock.acquire()
+            thread1.start()
+            sleep(1)
```

Comment:
Am I correct, that you need the sleep, so you are sure, that thread 1 reads first, before theard 2 writes new data?
Maybe you can use an [event](https://docs.python.org/3/library/threading.html#event-objects) here: The event is set in the mocked datastore write method just before acquiring the lock. The main thread can wait for the event and start thread 2.

This would be the execution order:
- main thread: acquire the lock, so thread 1 is blocked
- main thread: start thread 1 and wait for the event
- thread 1: fetches data, set the event, wait for the lock
- main thread: get the event, so thread 2 can be started
- main thread: wait for thread 2 to finish, release lock
- thread 1 finishes with some error

Or with the retires: no error. Maybe it should be asserted, that a retry happened.

This would require to add another element next to the lock, but since you have the structure already it should be fairly simple.

Reply:
Done